support-workers/src/main/scala/com/gu/zuora/subscriptionBuilders/DigitalSubscriptionPurchaseBuilder.scala

```diff
@@ -2,81 +2,63 @@ package com.gu.zuora.subscriptionBuilders
 
 import java.util.UUID
 
-import cats.data.EitherT
-import cats.implicits._
 import com.gu.i18n.Country
-import com.gu.monitoring.SafeLogger
 import com.gu.support.config.{TouchPointEnvironment, ZuoraDigitalPackConfig}
 import com.gu.support.promotions.{PromoCode, PromoError, PromotionService}
 import com.gu.support.redemption.gifting.GiftCodeValidator
-import com.gu.support.redemption.gifting.generator.CodeBuilder.GiftCode
-import com.gu.support.redemption.gifting.generator.GiftCodeGeneratorService
 import com.gu.support.workers.ProductTypeRatePlans.digitalRatePlan
-import com.gu.support.workers.{BillingPeriod, DigitalPack}
+import com.gu.support.workers.{DigitalPack, GeneratedGiftCode}
 import com.gu.support.zuora.api.ReaderType.Gift
-import com.gu.support.zuora.api.{ReaderType, SubscriptionData}
-import com.gu.zuora.subscriptionBuilders.ProductSubscriptionBuilders.{applyPromoCodeIfPresent, buildProductSubscription, validateRatePlan}
+import com.gu.support.zuora.api._
+import com.gu.zuora.subscriptionBuilders.ProductSubscriptionBuilders.{applyPromoCodeIfPresent, validateRatePlan}
 import org.joda.time.LocalDate
 
-import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.ExecutionContext
 
 class DigitalSubscriptionPurchaseBuilder(
   config: ZuoraDigitalPackConfig,
   promotionService: PromotionService,
-  giftCodeGenerator: GiftCodeGeneratorService,
   today: () => LocalDate,
 ) {
 
-  import DigitalSubscriptionPurchaseBuilder.addRedemptionCodeIfGift
-
   def build(
     maybePromoCode: Option[PromoCode],
-    billingPeriod: BillingPeriod,
     country: Country,
     digitalPack: DigitalPack,
     requestId: UUID,
     environment: TouchPointEnvironment,
-  )(implicit ec: ExecutionContext): EitherT[Future, PromoError, SubscriptionData] = {
+    maybeGiftCode: Option[GeneratedGiftCode],
+  )(implicit ec: ExecutionContext): Either[PromoError, SubscriptionData] = {
 
     val productRatePlanId = validateRatePlan(digitalRatePlan(digitalPack, environment), digitalPack.describe)
-    val (contractAcceptanceDelay, autoRenew, initialTerm) =
-      if (digitalPack.readerType == ReaderType.Gift)
-        (0, false, GiftCodeValidator.expirationTimeInMonths + 1)
-      else
-        (config.defaultFreeTrialPeriod + config.paymentGracePeriod, true, 12)
+
+    val (contractAcceptanceDelay, autoRenew, initialTerm, maybeRedemptionCode) =
+      (digitalPack.readerType, maybeGiftCode) match {
+        case (Gift, Some(giftCode)) => (0, false, GiftCodeValidator.expirationTimeInMonths + 1, Some(giftCode))
+        case (Gift, _) | (_, Some(_)) => throw new RuntimeException("coding error - possible unredeemable sub")
```

Comment:
we need to explode at this point - the types we have defined are not good enough to make this impossible, but it's impossible